pinot-common/src/main/java/org/apache/pinot/common/function/sql/PinotFunctionRegistry.java

```diff
@@ -0,0 +1,237 @@
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.schema.Function;
+import org.apache.calcite.sql.SqlIdentifier;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.SqlOperatorTable;
+import org.apache.calcite.sql.SqlSyntax;
+import org.apache.calcite.sql.SqlUtil;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.validate.SqlNameMatchers;
+import org.apache.calcite.sql.validate.SqlUserDefinedFunction;
+import org.apache.calcite.util.NameMultimap;
+import org.apache.pinot.common.function.TransformFunctionType;
+import org.apache.pinot.common.function.registry.PinotFunction;
+import org.apache.pinot.common.function.registry.PinotScalarFunction;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.segment.spi.AggregationFunctionType;
+import org.apache.pinot.spi.annotations.ScalarFunction;
+import org.apache.pinot.spi.utils.PinotReflectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Registry for scalar functions.
+ */
+public class PinotFunctionRegistry {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PinotFunctionRegistry.class);
+  private static final NameMultimap<SqlOperator> OPERATOR_MAP = new NameMultimap<>();
+  private static final NameMultimap<PinotFunction> FUNCTION_MAP = new NameMultimap<>();
+
+  private PinotFunctionRegistry() {
+  }
+
+  /**
+   * Registers the scalar functions via reflection.
+   * NOTE: In order to plugin methods using reflection, the methods should be inside a class that includes ".function."
+   *       in its class path. This convention can significantly reduce the time of class scanning.
+   */
+  static {
+    // REGISTER FUNCTIONS
+    long startTimeMs = System.currentTimeMillis();
+    Set<Method> methods = PinotReflectionUtils.getMethodsThroughReflection(".*\\.function\\..*", ScalarFunction.class);
+    for (Method method : methods) {
+      if (!Modifier.isPublic(method.getModifiers())) {
+        continue;
+      }
+      ScalarFunction scalarFunction = method.getAnnotation(ScalarFunction.class);
+      if (scalarFunction.enabled()) {
+        // Parse annotated function names and alias
+        Set<String> scalarFunctionNames = Arrays.stream(scalarFunction.names()).collect(Collectors.toSet());
+        if (scalarFunctionNames.size() == 0) {
+          scalarFunctionNames.add(method.getName());
+        }
+        boolean nullableParameters = scalarFunction.nullableParameters();
+        PinotFunctionRegistry.registerFunction(method, scalarFunctionNames, nullableParameters);
+      }
+    }
+    LOGGER.info("Initialized FunctionRegistry with {} functions: {} in {}ms", FUNCTION_MAP.map().size(),
+        FUNCTION_MAP.map().keySet(), System.currentTimeMillis() - startTimeMs);
+
+    // REGISTER OPERATORS
```

Comment:
Note to reviewer: 
this part will be maintain until all Transform function are registered in scalar form via PinotScalarFunction